File: Travelhub/Travelhub.swift/Models/RouteLocation.swift
//
//  RouteLocation.swift
//  Travelhub
//
//  Created by Jonas Groll on 11.02.26.
//

import Foundation
import SwiftData

@Model
final class RouteLocation {
    var id: String = UUID().uuidString
    var tripId: String
    var name: String
    var latitude: Double = 0
    var longitude: Double = 0
    var visitDate: Date
    var duration: Int = 1 // in days
    var notes: String = ""
    var order: Int = 0
    
    init(tripId: String, name: String, visitDate: Date) {
        self.tripId = tripId
        self.name = name
        self.visitDate = visitDate
    }
}

File: travelhub_flutter/lib/models/route_location.dart
import 'package:uuid/uuid.dart';

const _uuid = Uuid();

class RouteLocation {
  final String id;
  String tripId;
  String name;
  double latitude;
  double longitude;
  DateTime visitDate;
  int duration; // in days
  String notes;
  int order;

  RouteLocation({
    String? id,
    required this.tripId,
    required this.name,
    required this.visitDate,
    this.latitude = 0,
    this.longitude = 0,
    this.duration = 1,
    this.notes = '',
    this.order = 0,
  }) : id = id ?? _uuid.v4();

  Map<String, dynamic> toMap() => {
        'id': id,
        'tripId': tripId,
        'name': name,
        'latitude': latitude,
        'longitude': longitude,
        'visitDate': visitDate.toIso8601String(),
        'duration': duration,
        'notes': notes,
        'orderIndex': order,
      };

  factory RouteLocation.fromMap(Map<String, dynamic> m) => RouteLocation(
        id: m['id'] as String,
        tripId: m['tripId'] as String,
        name: m['name'] as String,
        visitDate: DateTime.parse(m['visitDate'] as String),
        latitude: (m['latitude'] as num?)?.toDouble() ?? 0,
        longitude: (m['longitude'] as num?)?.toDouble() ?? 0,
        duration: (m['duration'] as int?) ?? 1,
        notes: (m['notes'] ?? '') as String,
        order: (m['orderIndex'] as int?) ?? 0,
      );
}
